raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -109,51 +114,108 @@ private OptionalLong quorumMajorityFetchTimestamp() {
      *         not received fetch from the majority yet
      */
-    public OptionalLong updateFetchTimestamp(int nodeId, long timestamp) {
-        ReplicaState state = ensureValidVoter(nodeId);
-        // To be resilient to system time shifts we do not strictly require the timestamp be monotonically increasing
-        state.lastFetchTimestamp = OptionalLong.of(Math.max(state.lastFetchTimestamp.orElse(-1L), timestamp));
-        return quorumMajorityFetchTimestamp();
+    public OptionalLong updateReplicaFetchState(int nodeId,
+                                                long fetchOffset,
+                                                long leaderEndOffset,
+                                                long timestamp) {
+        ReplicaState state = getReplicaState(nodeId, leaderEndOffset);
+
+        state.updateFetchState(fetchOffset, leaderEndOffset, timestamp);
+
+        return isVoter(nodeId) ? quorumMajorityFetchTimestamp() : OptionalLong.empty();
     }
 
     public List<Integer> nonLeaderVotersByDescendingFetchOffset() {
         return followersByDescendingFetchOffset().stream()
                    .filter(state -> state.nodeId != localId)
-                   .map(state -> state.nodeId).collect(Collectors.toList());
+                   .map(state -> state.nodeId)
+                   .collect(Collectors.toList());
     }
 
-    private List<ReplicaState> followersByDescendingFetchOffset() {
-        List<ReplicaState> followersByDescendingFetchOffset = new ArrayList<>(this.voterReplicaStates.values());
-        Collections.sort(followersByDescendingFetchOffset);
-        return followersByDescendingFetchOffset;
+    private List<VoterState> followersByDescendingFetchOffset() {
+        return new ArrayList<>(this.voterReplicaStates.values())
+            .stream().sorted().collect(Collectors.toList());
     }
 
     /**
      * @return true if the high watermark is updated too
      */
     public boolean updateEndOffset(int remoteNodeId, LogOffsetMetadata endOffsetMetadata) {
-        ReplicaState state = ensureValidVoter(remoteNodeId);
+        // Ignore fetches from negative replica id, as it indicates
+        // the fetch is from non-replica. For example, a consumer.
+        if (remoteNodeId < 0) {
+            return false;
+        }
+
+        ReplicaState state = getReplicaState(remoteNodeId, endOffsetMetadata.offset);
+
         state.endOffset.ifPresent(currentEndOffset -> {
             if (currentEndOffset.offset > endOffsetMetadata.offset)
                 throw new IllegalArgumentException("Non-monotonic update to end offset for nodeId " + remoteNodeId);
         });
-        state.hasEndorsedLeader = true;
+
         state.endOffset = Optional.of(endOffsetMetadata);
-        return updateHighWatermark();
+
+        if (isVoter(remoteNodeId)) {
+            ((VoterState) state).hasEndorsedLeader = true;
```

Comment:
nit: we can use "ensureValidVoter" here.

Reply:
I feel it's a bit overkill, since we already checked `isVoter`

Reply:
SG. It is a rather small nit anyways, I was thinking we can wrap the casting to something like `VoterStateOrThrow`, but since we only do this casting once it's not a big deal.